pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +439,143 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+// when a deployment is paused and not during the normal rollout process.
+func (dc *DeploymentController) scale(deployment *extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) error {
+	// If there is only one active replica set then we should scale that up to the full count of the
+	// deployment. If there is no active replica set, then we should scale up the newest replica set.
+	if activeOrLatest := findActiveOrLatest(newRS, oldRSs); activeOrLatest != nil {
+		if activeOrLatest.Spec.Replicas == deployment.Spec.Replicas {
+			return nil
+		}
+		_, _, err := dc.scaleReplicaSetAndRecordEvent(activeOrLatest, deployment.Spec.Replicas, deployment)
+		return err
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// There are old replica sets with pods and the new replica set is not saturated.
+	// We need to proportionally scale all replica sets (new and old) in case of a
+	// rolling deployment.
+	if deploymentutil.IsRollingUpdate(deployment) {
+		allRSs := controller.FilterActiveReplicaSets(append(oldRSs, newRS))
+		allRSsReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+
+		// Number of additional replicas that can be either added or removed from the total
+		// replicas count. These replicas should be distributed proportionally to the active
+		// replica sets. Any leftovers should be distributed as well.
+		deploymentReplicasToAdd := deployment.Spec.Replicas - allRSsReplicas
+		if deploymentReplicasToAdd == 0 {
+			return nil
+		}
+
+		// The additional replicas should be distributed proportionally amongst the active
+		// replica sets from the larger to the smaller in size replica set. Scaling direction
+		// drives what happens in case we are trying to scale replica sets of the same size.
+		// In such a case when scaling up, we should scale up newer replica sets first, and
+		// when scaling down, we should scale older replica sets first.
+		scaleUp := deploymentReplicasToAdd > 0
+		scalingOperation := "up"
+		if scaleUp {
+			sort.Sort(controller.ReplicaSetsBySizeNewer(allRSs))
+		} else {
+			sort.Sort(controller.ReplicaSetsBySizeOlder(allRSs))
+			scalingOperation = "down"
+		}
+
+		// Iterate over all active replica sets and estimate proportions for each of them.
+		// The absolute value of deploymentReplicasAdded should never exceed the absolute
+		// value of deploymentReplicasToAdd.
+		deploymentReplicasAdded := int32(0)
+		for i := range allRSs {
+			rs := allRSs[i]
+
+			proportion := getProportion(rs, deployment.Spec.Replicas, deploymentReplicasToAdd, deploymentReplicasAdded)
```

Comment:
This is looking really close.  I believe I thought of an edge case this morning when using proportional scaling.  There are a couple ways we can handle this.

- Deployment count = 10 replicas
- MaxSurge = 2 replicas
- RS1 = 5 + 1surge replicas
- RS2 = 5 + 1surge replicas

Scale to 20 replicas
- RS1 = 20/10*6 = 12
- RS2 = 20/10*6 = 12
Total: 24 != desired + max-surge = 22

I think the left over logic may even exacerbate the issue by adding abs(-2) left over replicas.

There are a couple ways we can handle this:
1. Keep track of replicas and stop when we hit the limit - portions may be a bit off, but not terribly
2. Keep track of total replicas added and normalize each portion by multiplying it against (total we would add / total we want to add).  Still probably want to keep track of limits as in 1 as a sanity check.
3. Use the other algorithm suggested.

Right now I am thinking 2 is the most elegant and pretty close to what you have written.  WDYT?

Overall this is a big improvement over what we have, and we just need to make sure its behavior is sane and predictable.  (Even if it is better scaling, I think it would be worse overall if it was unpredictable in how it scaled things)